pkg/monitor/cluster/podconditions.go

```diff
@@ -120,3 +123,33 @@ func (mon *Monitor) _emitPodContainerStatuses(ps *corev1.PodList) {
 		}
 	}
 }
+
+func (mon *Monitor) _emitPodContainerRestartCounter(ps *corev1.PodList) {
+	for _, p := range ps.Items {
+		if !namespace.IsOpenShiftSystemNamespace(p.Namespace) {
+			continue
+		}
+
+		for _, cs := range p.Status.ContainerStatuses {
+			if cs.RestartCount < restartCounterThreadshold {
+				continue
+			}
```

Comment:
Maybe we should always send metrics and deal with the alerting logic on the monitor side? That way we can see the full picture somewhere on the dashbaord: not only starting after N restarts. And we will be able to adjust threshold without an RP release. I see that in the [work iterm](https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/9534801) it was suggested to have this condition in code, but it is not super clear why. Maybe some Geneva limits? @gvanderpotte (and any one else) any thougts?

Also wondering - what is the plan for alerting logic? Create an alert once we have more than N restarts? If so - do you know whether restart count resets at some point? Or is it just being incremented? If I recall correctly, it resets after some time of successfull execution, but I'm not 100% sure. After a quick search I managed to find [this](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy):

> Once a container has executed for 10 minutes without any problems, the kubelet resets the restart backoff timer for that container.

But it is about backoff timer, not about the counter.

Reply:
The reason for the threshold is that there ussually is a steady number of restarts after cluster deploy/upgrade. The reason for this is simply the design of the cluster and the eventual consistency. We have often observed around 4 restarts for e.g. for network before it went OK. 

Small amount of restarts ussually does not mean a problem, that was the point. 

As for me, I am OK with the threshold as long as it is kept low (like 5) to simply filter the bootstrapping restarts.

Reply:
@petrkotas yes, it makes sense to have a threshold. The question is - do we put the threshold filtering in the RP (before sending the metrics to Geneva) or in the monitor (when we have metrics in geneva already).

Having threshold sounds like more flexible option: we can adjust threshold without having to release an RP. But maybe there are some implications that I'm missing. What do you think?

Reply:
IMO it depends on the Threshold value. 10 seems ok to me.

If we set it too high - there will be a mismatch of expectations for someone looking at the dashboards when they see no data for pods that don't restart, when they're looking for one restart.

If we set it too low - we're going to get a ton of datapoints that aren't useful (I assume 2-3 restarts on a pod are within normal execution flow).

Reply:
There was also the point of using a threshold is/was to save on Geneva data costs. 

I don't think the restart counter itself gets reset during the lifetime of a pod (or even beyond https://github.com/kubernetes/kubernetes/issues/54476) . There is a mechanism that pauses restarting after a number of failed attempts in a short amount of time  for some time. 

Having hard-coded values in the metric generation code  does not feel right , but that seems to be the design followed for other metrics, too.

Reply:
> Having hard-coded values in the metric generation code does not feel right , but that seems to be the design followed for other metrics, too.

On the metric generation side we normally filter out things like healthy conditions: we only send unhealthy/unexpected conditions. But I don't remember having things like thresholds hardcoded anywhere.

I do not have strong feelings about it: I'm ok with having it here. Having threshold in the monitor side means we send more data, but we are more flexible with the thresholds value (no need to redelpoy RP if we want to tweak it).

However all this probably doesn't really matter on the bigger picture becuase of this:


> I don't think the restart counter itself gets reset during the lifetime of a pod (or even beyond kubernetes/kubernetes#54476) . There is a mechanism that pauses restarting after a number of failed attempts in a short amount of time for some time.

In that case we might have a issue: at what point do we start sending metrics (and more importantly start alerting)? If the restart counter is not being reset - this means that we probably already have some older clusters which are currently healthy but have more than 10 restarts on the counter.

Reply:
The way I understand the ticket, we'd alert on change rate, not total number of restarts.

Afaic we have two phases:
- Data collection: Here the threshold of 10 (or similar) helps reducing the total amount of data we send into Geneva, because we filter out all of the healthy pods on the  RP side.
- Data evaluation: Here we'd look at the rate with which the restarts occur and alert if that rate exceeds another threshold (that we define in Geneva/the monitor). The absolute restart counts would not matter in this case.

I have no good data to suggest a good value for the rate at this point in time. During my tests I did not manage to get above something like 5-10 restarts per hour because of the mentioned protection mechanism, which just makes the cluster stop restarting.

Reply:
> Data collection: Here the threshold of 10 (or similar) helps reducing the total amount of data we send into Geneva, because we filter out all of the healthy pods on the RP side.

It becomes less useful given that restart counters do not get reset. But I'm ok with that.

> Data evaluation: Here we'd look at the rate with which the restarts occur and alert if that rate exceeds another threshold (that we define in Geneva/the monitor). The absolute restart counts would not matter in this case.

This makes. Did you have any success defining alerting logic like this in Geneva?